protocols/abci/addressing-1174.md

```diff
@@ -0,0 +1,203 @@
+implementation of `PrepareProposal` and `ProcessProposal` fulfills them.
+All applications that are able to enforce these properties do not need to reason about
+the internals of the consensus implementation: they can consider it as a black box.
+This is the most desirable situation in terms of modularity between CometBFT and the application.
+
+## Problem Statement
+
+This document is dealing with the case when an application cannot guarantee
+the coherence and/or determinism requirements as stated in the ABCI 1.0 specification.
+
+An example of this is when `ProcessProposal` needs to take inputs from third-party entities
+(e.g. price oracles) that are not guaranteed to provide exactly the same values to
+different processes during the same height.
+Another example is when `ProcessProposal` needs to read the system clock in order to perform its checks
+(e.g. Proposer-Based Timestamp when expressed as an ABCI 1.0 application).
+
+In principle, if an application's implementation of `PrepareProposal` and `ProcessProposal`
+is not able to fulfill coherence and determinism requirements,
+CometBFT cannot guarantee consensus _termination_ in all runs of the system.
+As a result, the application designers must start considering both CometBFT and their application
+as one monolithic block, in order to reason about termination.
+We thus lose the modularity provided when fulfilling the ABCI 1.0 requirements.
+Remember that CometBFT's consensus algorithm (Tendermint) is a well-known algorithm that
+has been studied, reviewed, formally analyzed, model-checked, etc.
+The combination of CometBFT and an arbitrary application as one single algorithm cannot
+leverage that extensive body of research applied to the Tendermint algorithm.
+This situation is risky and undesirable.
+
+So, the questions that arises are the following.
+Can we come up with a set of weaker requirements
+that applications unable to fulfill the current ABCI 1.0 requirements
+can still fulfill?
+Is this set of weaker requirements still strong enough to guarantee consensus _termination_?
+
+## Solution Proposed
+
+### Modified consensus _validity_
+The specification of consensus implicitly considers `valid(v)` as an immutable function,
+which is always supposed to provide the same result, when called at the same height
+for the same value `v`.
+This was the main reason for introducing the determinism requirements on `ProcessProposal`
+in the ABCI 1.0 specification.
+
+If we are to relax the determinism requirements on the application,
+we first need to weaken the `valid(v)` predicate and therefore consensus _validity_.
+In the proposed solution, correct processes MAY evaluate `valid(v)` to _true_ or _false_ at different times
+for the same value `v` during height `h`.
+
+Consensus _validity_ invariant is then modified as follows:
+
+- _weak validity_: predicate `valid(v)` has been evaluated to _true_ at least once by one correct process
+  for the decided block `v`.
```

Comment:
I think this write-up is a bit imprecise. It mixes "mathematical functions" (as valid) and code functions (as prepareproposal).

Reply:
OK, I'll see what I can do

Reply:
Let me know if it looks better now

Reply:
Yes, I agree. If at least one correct processes says a value is valid, this requires $f+1$ processes saying it is valid.

Reply:
Great, thanks!